src/solvers/dg_structured/dg_3d.jl

```diff
@@ -0,0 +1,293 @@
+    element = cache.elements[element_ind]
+
+    # Interfaces in x-direction
+    for k in eachnode(dg), j in eachnode(dg), v in eachvariable(equations)
+      element.interfaces[1].u_right[v, j, k] = u[v, 1, j, k, element_ind]
+      element.interfaces[2].u_left[v, j, k] = u[v, end, j, k, element_ind]
+    end
+
+    # Interfaces in y-direction
+    for k in eachnode(dg), i in eachnode(dg), v in eachvariable(equations)
+      element.interfaces[3].u_right[v, i, k] = u[v, i, 1, k, element_ind]
+      element.interfaces[4].u_left[v, i, k] = u[v, i, end, k, element_ind]
+    end
+
+    # Interfaces in z-direction
+    for j in eachnode(dg), i in eachnode(dg), v in eachvariable(equations)
+      element.interfaces[5].u_right[v, i, j] = u[v, i, j, 1, element_ind]
+      element.interfaces[6].u_left[v, i, j] = u[v, i, j, end, element_ind]
+    end
+  end
+
+  return nothing
+end
+
+
+function prolong2boundaries!(cache, u::AbstractArray{<:Any,5}, 
+    boundary_condition::BoundaryConditionPeriodic, mesh::StructuredMesh, equations, dg::DG)
+  @unpack size, linear_indices = mesh
+
+  # Boundaries in x-direction
+  for element_y in 1:size[2], element_z in 1:size[3]
+    # TODO Performance
+    cache.elements[1, element_y, element_z].interfaces[1].u_left .= u[:, end, :, :, linear_indices[end, element_y, element_z]]
+    cache.elements[end, element_y, element_z].interfaces[2].u_right .= u[:, 1, :, :, linear_indices[1, element_y, element_z]]
+  end
+
+  # Boundaries in y-direction
+  for element_x in 1:size[1], element_z in 1:size[3]
+    cache.elements[element_x, 1, element_z].interfaces[3].u_left .= u[:, :, end, :, linear_indices[element_x, end, element_z]]
+    cache.elements[element_x, end, element_z].interfaces[4].u_right .= u[:, :, 1, :, linear_indices[element_x, 1, element_z]]
+  end
+
+  # Boundaries in z-direction
+  for element_x in 1:size[1], element_y in 1:size[2]
+    cache.elements[element_x, element_y, 1].interfaces[5].u_left .= u[:, :, :, end, linear_indices[element_x, element_y, end]]
+    cache.elements[element_x, element_y, end].interfaces[6].u_right .= u[:, :, :, 1, linear_indices[element_x, element_y, 1]]
+  end
+
+  return nothing
+end
+
+
+function calc_interface_flux!(nonconservative_terms::Val{false}, mesh::StructuredMesh{3}, equations,
+                              dg::DG, cache)
+  @unpack surface_flux = dg
+  @unpack size = mesh
+
+  @threaded for element in eachelement(dg, cache)
+    # Interfaces in negative directions
+    for orientation in (1, 3, 5)
+      interface = cache.elements[element].interfaces[orientation]
+      calc_interface_flux!(interface, mesh, equations, dg)
+    end
+  end
+
+  # Boundary in positive x-direction
+  # TODO maybe make this threaded with @threaded for element in @view(mesh.linear_indices[end, :, :])
+  for element_y in 1:size[2], element_z in 1:size[3]
+    interface = cache.elements[end, element_y, element_z].interfaces[2]
+    calc_interface_flux!(interface, mesh, equations, dg)
+  end
```

Comment:
Running `examples/3d/elixir_euler_source_terms_structured.jl` with `8^3` cells led to these results:
```
 ───────────────────────────────────────────────────────────────────────────────
            Trixi.jl                    Time                   Allocations
                                ──────────────────────   ───────────────────────
        Tot / % measured:            6.48s / 87.9%           1.12GiB / 100%

 Section                ncalls     time   %tot     avg     alloc   %tot      avg
 ───────────────────────────────────────────────────────────────────────────────
 rhs!                    3.16k    4.68s  82.1%  1.48ms   1.07GiB  95.6%   354KiB
   interface flux        3.16k    1.45s  25.5%   460μs   29.2MiB  2.55%  9.46KiB
   prolong2boundaries    3.16k    749ms  13.1%   237μs    938MiB  82.0%   304KiB
   volume integral       3.16k    678ms  11.9%   214μs   29.9MiB  2.61%  9.67KiB
   prolong2interfaces    3.16k    522ms  9.15%   165μs   29.2MiB  2.55%  9.45KiB
   surface integral      3.16k    452ms  7.94%   143μs   29.4MiB  2.57%  9.53KiB
   source terms          3.16k    406ms  7.12%   128μs   33.4MiB  2.92%  10.8KiB
   Jacobian              3.16k    323ms  5.67%   102μs     32.0B  0.00%    0.01B
   reset ∂u/∂t           3.16k   71.4ms  1.25%  22.6μs     0.00B  0.00%    0.00B
 analyze solution            8    695ms  12.2%  86.9ms   11.6MiB  1.02%  1.45MiB
 calculate dt              633    229ms  4.02%   362μs     0.00B  0.00%    0.00B
 I/O                        17   97.7ms  1.71%  5.75ms   39.0MiB  3.41%  2.29MiB
 ───────────────────────────────────────────────────────────────────────────────
```
It turns out that the interface flux takes the most time (more than with a `TreeMesh`).
The first `@threaded` loop is fine, but the three other loops combined take the same time as the first one, while the first one computes 8 times as many interface fluxes.
We haven't come up with a way to make the three small loops threaded properly. With the line in the `TODO` comment, we could make it threaded, but it wasn't really faster.

What would be the best way to make this faster?

Reply:
Okay, so here the `calc_interface_flux!` is nearly twice as slow as with `TreeMesh`. As I said above, the `@threaded` loop takes about half of the time here, so this loop alone is about as fast as with `TreeMesh` (a bit faster even). I guess that the other three loops are slower only because they are not threaded.

We looked at this code again, and noticed that we only need to use the other three loops because we used different (but identical) interfaces for the left and right boundary, even though we use periodic boundaries. We changed this, so now we only have this one threaded loop that is a bit faster than with `TreeMesh`.

I'll comment the whole timer output below.

Reply:
Have you changed the data structures yet? If not, I personally wouldn't worry too much about the performance of this.

In general, serial performance benchmarking should *always* be conducted on a single thread, where our `@threaded` macro ensures that all threading macros disappear.